File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
from collections import defaultdict
adj = defaultdict(list)
ans = []


def dfs(cur):
    while adj[cur]:
        dfs(adj[cur].pop(0))
    if not adj[cur]:
        ans.append(cur)
        return


def solution(t):
    for a, b in t:
        adj[a].append(b)
    for value in adj.values():
        value.sort()
    dfs('ICN')
    return ans[::-1]

File: 프로그래머스/lv3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    int n = 0;
    Map<String, List<String>> adj = new HashMap<>();
	List<String> ans = new ArrayList<>();
    public String[] solution(String[][] t) {
        n = t.length;

        for (int i = 0; i < n; i++) {
            if (!adj.containsKey(t[i][0])) adj.put(t[i][0], new ArrayList<>());
			if (!adj.containsKey(t[i][1])) adj.put(t[i][1], new ArrayList<>());

			List<String> tmp = adj.get(t[i][0]);
			tmp.add(t[i][1]);
            adj.put(t[i][0], tmp);
        }

		for (String key: adj.keySet()) {
			Collections.sort(adj.get(key));
		}


		dfs("ICN");
		Collections.reverse(ans);

        return ans.toArray(new String[0]);
    }

	void dfs(String cur) {
		while (!adj.get(cur).isEmpty()) {
			dfs(adj.get(cur).remove(0));
		}

		if (adj.get(cur).isEmpty()) {
			ans.add(cur);
			return;
		}
	}
}